pkg/receive/handler.go

```diff
@@ -147,6 +156,13 @@ func NewHandler(logger log.Logger, o *Options) *Handler {
 			Max:    30 * time.Second,
 			Jitter: true,
 		},
+		writeGate: gate.NewNoop(),
+		requestLimiter: newRequestLimiter(
```

Comment:
Instead of extending the existing handler, is it possible to make the limiter a middleware that wraps the handler? This could be cleaner to maintain long term.

Reply:
Unfortunately I don't think that's a good idea, because then the limiter middleware will have to do work that is done again in the main handler: 

- Checking content length header is easy, but when it's not there we have to read the request body to check its size, which involves trying a snappy decompression too.
- Unmarshalling the request body and looping over the data to count timeseries and samples.

I wanted to avoid doing extra work at all costs to keep the "hot path" of Receive fast.

Reply:
I see, that's a good point. Well initially I thought the middleware would be on [this](https://github.com/douglascamata/thanos/blob/7efee5e7d59527c7464a84b1b2d9436691ed74bf/pkg/receive/handler.go#L370) handler which already has the decoded protobuf request, but here we don't have the raw body correct?

Also if I may ask, what is the practical use of setting limits on the request body? As an admin, I am not sure I would know what to set this limit to. I can understand the impact of limiting series or samples, but I would find it hard to know how to configure the body size limit.